psh/dd/dd.c

```diff
@@ -0,0 +1,438 @@
+	param_conv, param_count, param_seek, param_skip };
+
+
+static const struct param_s params[] = {
+	{ "if", param_if },
+	{ "of", param_of },
+	{ "bs", param_bs },
+	{ "conv", param_conv },
+	{ "count", param_count },
+	{ "seek", param_seek },
+	{ "skip", param_skip },
+	{ NULL, param_none }
+};
+
+
+enum { conv_none = 0, conv_nocreat, conv_notrunc };
+
+
+static const struct param_s convs[] = {
+	{ "nocreat", conv_nocreat },
+	{ "notrunc", conv_notrunc },
+	{ NULL, conv_none }
+};
+
+/* clang-format on */
+
+
+static void psh_dd_info(void)
+{
+	printf("copy a file according to the operands");
+}
+
+
+static void psh_dd_usage(void)
+{
+	printf(
+		"Usage: dd [OPERAND]...\n"
+		"\tif=FILE     read from FILE instead of stdin\n"
+		"\tof=FILE     write to FILE instead of stdout\n"
+		"\tbs=BYTE     read/write block size of BYTES bytes\n"
+		"\tcount=N     copy only N input blocks\n"
+		"\tseek=N      skip N bs-sized blocks at start of output\n"
+		"\tskip=N      skip N bs-sized blocks at start of input\n"
+		"\tconv=CONVS  comma-separated list of supported conversions:\n"
+		"\t            e.g. nocreat,notrunc\n");
+}
+
+
+static ssize_t getnumber(const char *numstr)
+{
+	char *str;
+	volatile ssize_t retval;
+	volatile ssize_t mult = 1;
+	volatile ssize_t value = strtoul(numstr, &str, 10);
+
+	if (numstr == str) {
+		return -EINVAL;
+	}
+
+	switch (*(str++)) {
+		case 'M':
+			mult = 1024 * 1024;
+			break;
+
+		case 'k':
+			mult = 1024;
+			break;
+
+		case 'b':
+			mult = 512;
+			break;
+
+		case 'w':
+			mult = 2;
+			break;
+
+		case 'c':
+			/* mult *= 1; */
+			break;
+
+		case '\0':
+			return value;
+
+		default:
+			return -1;
+	}
+
+	if (*str != '\0') {
+		return -EINVAL;
+	}
+
+	retval = value * mult;
+	if (value != (retval / mult)) {
+		return -EOVERFLOW;
+	}
+
+	return retval;
+}
+
+
+static int getconvmode(mode_t *m, const char *str)
+{
+	const char *end;
+	const struct param_s *par;
+	mode_t mode = *m;
+
+	for (end = str; *end != '\0'; str = end + 1) {
+		end = strchr(str, ',');
+		if (end == NULL) {
+			end = str + strlen(str);
+		}
+
+		for (par = convs; par->name != NULL; par++) {
+			size_t len = end - str;
+			if ((strncmp(str, par->name, len) == 0) && (par->name[len] == '\0')) {
+				break;
+			}
+		}
+
+		switch (par->value) {
+			case conv_nocreat:
+				mode &= ~O_CREAT;
+				break;
+
+			case conv_notrunc:
+				mode &= ~O_TRUNC;
+				break;
+
+			default:
+				return -EINVAL;
+		}
+	}
+
+	*m = mode;
+
+	return EOK;
+}
+
+
+static int psh_dd(int argc, char **argv)
+{
+	const struct param_s *par;
+	const char *str, *cp;
+	char *buf, *p;
+	int err = 0, infd, outfd;
+	ssize_t incc, outcc;
+	ssize_t intotal, outtotal;
+
+	const char *infile = NULL;
+	const char *outfile = NULL;
+	mode_t outmode = O_CREAT | O_TRUNC;
+	volatile ssize_t tmp, count = SSIZE_MAX;
+	volatile ssize_t seekval = 0;
+	volatile ssize_t skipval = 0;
+	volatile ssize_t inmax = 0;
+	volatile int blocksz = 512;
```

Comment:
Hmm, volatiles shouldn't be needed, this is one thread app, no interrupts/signal either. Why add them?

Reply:
While testing of `dd`, I found that without `volatile` the overflow checks (below) are removed by the gcc optimizer.
```c
	retval = value * mult;
	if (value != (retval / mult)) {
		return -EOVERFLOW;
	}
```
and
```c
	tmp = skipval;
	skipval *= blocksz;
	if (tmp != skipval / blocksz) {
	...

	tmp = seekval;
	seekval *= blocksz;
	if (tmp != seekval / blocksz) {
	...

	inmax = count * blocksz;
	if (count != inmax / blocksz) {   
	...
```